# Roundtrip Fidelity Benchmark

Measures how faithfully AIF documents survive a roundtrip through each format:

**AIF -> Format (HTML, Markdown, JSON) -> Re-import -> Compare IR**

**Why this matters for LLM understanding:** Typed semantic blocks (`@claim`, `@evidence`,
`@definition`, `@step`, `@verify`) give LLMs explicit structural cues for reasoning.
When a roundtrip degrades these blocks into flat text, the LLM loses its ability to
distinguish factual claims from supporting evidence, instructions from commentary.
This benchmark measures whether each format preserves the structural signals that
help LLMs understand and follow document content accurately.

In [1]:
import json
import os
import subprocess
import sys
import tempfile
import statistics
from pathlib import Path
from collections import Counter

PROJECT_ROOT = Path(os.getcwd()).resolve()
# Walk up until we find Cargo.toml (workspace root)
while not (PROJECT_ROOT / "Cargo.toml").exists() and PROJECT_ROOT != PROJECT_ROOT.parent:
    PROJECT_ROOT = PROJECT_ROOT.parent

EXAMPLES_DIR = PROJECT_ROOT / "examples"
FORMATS = ["html", "markdown", "json"]

print(f"Project root: {PROJECT_ROOT}")
print(f"Examples dir: {EXAMPLES_DIR}")
print(f"Formats: {', '.join(FORMATS)}")

Project root: /Users/liqunchen/Claude_Project/token_efficient
Examples dir: /Users/liqunchen/Claude_Project/token_efficient/examples
Formats: html, markdown, json


## Methodology

For each `.aif` file in the examples directory, the benchmark performs a three-step roundtrip
through each target format:

1. **Dump original IR** -- parse the `.aif` file into JSON IR via `aif dump-ir`
2. **Compile to format** -- emit HTML, Markdown, or JSON via `aif compile`
3. **Re-import** -- import the compiled output back to JSON IR via `aif import` (or JSON re-compile)
4. **Compare** -- diff the original and roundtripped IRs across 6 fidelity metrics

The comparison is purely structural: we compare block trees, semantic types, metadata keys,
and inline element distributions between the original and roundtripped documents.

## Metric Definitions

| Metric | Weight | Description |
|--------|--------|-------------|
| **Block count ratio** | (informational) | `roundtripped_blocks / original_blocks` -- 1.0 means same count |
| **Block type preservation** | 1x | Fraction of blocks that keep the same `BlockKind` type in position-aligned comparison |
| **Semantic type preservation** | **2x** | Fraction of `SemanticBlock` types (claim, evidence, definition, etc.) that survive the roundtrip. Double-weighted because this is the core AIF value proposition |
| **Metadata preservation** | 1x | Fraction of original metadata keys preserved (excluding `_aif_*` provenance keys) |
| **Inline fidelity** | 1x | Fraction of inline element types (emphasis, strong, code, links) preserved by multiset matching |
| **Overall fidelity** | -- | Weighted average: `(block_types*1 + semantic*2 + metadata*1 + inlines*1) / 5` |

In [2]:
# ── CLI helpers ──────────────────────────────────────────────────────────────

def run_cli(args: list, stdin_data=None) -> str:
    """Run aif-cli and return stdout."""
    result = subprocess.run(
        ["cargo", "run", "-p", "aif-cli", "--"] + args,
        capture_output=True, text=True, cwd=str(PROJECT_ROOT),
        input=stdin_data,
    )
    if result.returncode != 0:
        print(f"  CLI error ({' '.join(args[:3])}): {result.stderr.strip()[:200]}",
              file=sys.stderr)
        return ""
    return result.stdout


def dump_ir(aif_file: str):
    """Parse an .aif file and return its JSON IR."""
    out = run_cli(["dump-ir", aif_file])
    if not out:
        return None
    return json.loads(out)


def compile_to(aif_file: str, fmt: str, output_path: str) -> bool:
    """Compile .aif to a given format, writing to output_path."""
    run_cli(["compile", aif_file, "--format", fmt, "-o", output_path])
    return os.path.exists(output_path)


def import_file(file_path: str):
    """Import a file (html/md) back to JSON IR via stdout."""
    out = run_cli(["import", file_path])
    if not out:
        return None
    return json.loads(out)


def json_roundtrip(json_path: str):
    """Compile JSON IR back through the JSON format path."""
    out = run_cli(["compile", json_path, "--input-format", "json", "--format", "json"])
    if not out:
        return None
    return json.loads(out)


# ── Block / Inline extraction ────────────────────────────────────────────────

def collect_blocks(blocks: list) -> list:
    """Recursively collect all blocks from nested structures."""
    result = []
    for block in blocks:
        kind = block.get("kind", {})
        result.append(kind)
        block_type = kind.get("type", "")
        if block_type == "Section":
            result.extend(collect_blocks(kind.get("children", [])))
        elif block_type == "BlockQuote":
            result.extend(collect_blocks(kind.get("content", [])))
        elif block_type == "SkillBlock":
            result.extend(collect_blocks(kind.get("children", [])))
        elif block_type == "List":
            for item in kind.get("items", []):
                result.extend(collect_blocks(item.get("children", [])))
    return result


def collect_inlines(obj) -> list:
    """Recursively collect all inline type names from any JSON structure."""
    types = []
    if isinstance(obj, dict):
        if "type" in obj:
            t = obj["type"]
            if t in ("Text", "Emphasis", "Strong", "InlineCode", "Link",
                      "Image", "Reference", "Footnote", "SoftBreak", "HardBreak"):
                types.append(t)
        for v in obj.values():
            types.extend(collect_inlines(v))
    elif isinstance(obj, list):
        for item in obj:
            types.extend(collect_inlines(item))
    return types


def get_block_types(blocks: list) -> list:
    return [b.get("type", "Unknown") for b in blocks]


def get_semantic_types(blocks: list) -> list:
    return [b.get("block_type", "Unknown") for b in blocks if b.get("type") == "SemanticBlock"]


def get_metadata_keys(doc: dict) -> set:
    meta = doc.get("metadata", {})
    return {k for k in meta.keys() if not k.startswith("_aif_")}


# ── Comparison ───────────────────────────────────────────────────────────────

def compare_ir(original: dict, roundtripped: dict) -> dict:
    """Compare original and roundtripped IR, returning fidelity metrics."""
    orig_blocks = collect_blocks(original.get("blocks", []))
    rt_blocks = collect_blocks(roundtripped.get("blocks", []))

    orig_count = len(orig_blocks)
    rt_count = len(rt_blocks)

    # Block count ratio
    block_count_ratio = rt_count / orig_count if orig_count > 0 else 0.0

    # Block type preservation (position-aligned)
    orig_types = get_block_types(orig_blocks)
    rt_types = get_block_types(rt_blocks)
    matched_types = sum(1 for i, t in enumerate(orig_types) if i < len(rt_types) and rt_types[i] == t)
    block_type_pres = matched_types / len(orig_types) if orig_types else 1.0

    # Semantic type preservation (multiset matching)
    orig_sem = get_semantic_types(orig_blocks)
    rt_sem = get_semantic_types(rt_blocks)
    if orig_sem:
        orig_counter = Counter(orig_sem)
        rt_counter = Counter(rt_sem)
        preserved = sum(min(orig_counter[k], rt_counter.get(k, 0)) for k in orig_counter)
        semantic_pres = preserved / len(orig_sem)
    else:
        semantic_pres = 1.0

    # Metadata preservation
    orig_meta = get_metadata_keys(original)
    rt_meta = get_metadata_keys(roundtripped)
    meta_pres = len(orig_meta & rt_meta) / len(orig_meta) if orig_meta else 1.0

    # Inline fidelity (multiset matching)
    orig_inlines = collect_inlines(original.get("blocks", []))
    rt_inlines = collect_inlines(roundtripped.get("blocks", []))
    if orig_inlines:
        orig_icnt = Counter(orig_inlines)
        rt_icnt = Counter(rt_inlines)
        total_orig = sum(orig_icnt.values())
        preserved_inlines = sum(min(orig_icnt[k], rt_icnt.get(k, 0)) for k in orig_icnt)
        inline_fidelity = preserved_inlines / total_orig
    else:
        inline_fidelity = 1.0

    # Overall fidelity: weighted average (semantic 2x)
    overall = (block_type_pres * 1 + semantic_pres * 2 + meta_pres * 1 + inline_fidelity * 1) / 5.0

    return {
        "block_count_original": orig_count,
        "block_count_roundtripped": rt_count,
        "block_count_ratio": round(block_count_ratio, 4),
        "block_type_preservation": round(block_type_pres, 4),
        "semantic_type_preservation": round(semantic_pres, 4),
        "metadata_preservation": round(meta_pres, 4),
        "inline_fidelity": round(inline_fidelity, 4),
        "overall_fidelity": round(overall, 4),
    }


print("Benchmark functions loaded.")

Benchmark functions loaded.


In [3]:
# ── Run Benchmark ────────────────────────────────────────────────────────────
# This is DETERMINISTIC -- always runs fresh against the current codebase.

aif_files = sorted(EXAMPLES_DIR.glob("**/*.aif"))
print(f"Found {len(aif_files)} .aif files\n")

results = []

for aif_file in aif_files:
    filename = aif_file.name
    print(f"{'='*60}")
    print(f"  {filename}")
    print(f"{'='*60}")

    # Step 1: dump original IR
    original_ir = dump_ir(str(aif_file))
    if original_ir is None:
        print(f"  SKIP: could not parse {filename}")
        continue

    for fmt in FORMATS:
        print(f"  [{fmt:>10}] ", end="", flush=True)

        with tempfile.TemporaryDirectory() as tmpdir:
            ext = {"html": "html", "markdown": "md", "json": "json"}[fmt]
            compiled_path = os.path.join(tmpdir, f"compiled.{ext}")

            ok = compile_to(str(aif_file), fmt, compiled_path)
            if not ok:
                print("FAIL (compile)")
                results.append({"file": filename, "format": fmt, "error": "compile_failed"})
                continue

            if fmt == "json":
                rt_ir = json_roundtrip(compiled_path)
            else:
                rt_ir = import_file(compiled_path)

            if rt_ir is None:
                print("FAIL (import)")
                results.append({"file": filename, "format": fmt, "error": "import_failed"})
                continue

            metrics = compare_ir(original_ir, rt_ir)
            metrics["file"] = filename
            metrics["format"] = fmt
            results.append(metrics)

            print(f"blocks={metrics['block_count_ratio']:.2f}  "
                  f"types={metrics['block_type_preservation']:.2f}  "
                  f"semantic={metrics['semantic_type_preservation']:.2f}  "
                  f"meta={metrics['metadata_preservation']:.2f}  "
                  f"inline={metrics['inline_fidelity']:.2f}  "
                  f"overall={metrics['overall_fidelity']:.2f}")

print(f"\nCompleted: {len(results)} roundtrip tests")

Found 40 .aif files

  simple.aif


  [      html] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=1.00  overall=1.00
  [  markdown] 

blocks=1.27  types=0.09  semantic=0.00  meta=0.50  inline=1.00  overall=0.32
  [      json] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=1.00  overall=1.00
  wiki_article.aif


  [      html] 

blocks=1.00  types=1.00  semantic=1.00  meta=0.67  inline=1.00  overall=0.93
  [  markdown] 

blocks=1.30  types=0.22  semantic=0.00  meta=0.33  inline=1.00  overall=0.31
  [      json] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=1.00  overall=1.00
  migration_eslint_flat_config.aif


  [      html] 

blocks=1.00  types=1.00  semantic=1.00  meta=0.50  inline=0.67  overall=0.83
  [  markdown] 

blocks=1.93  types=0.00  semantic=1.00  meta=0.50  inline=0.97  overall=0.69
  [      json] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=1.00  overall=1.00
  migration_nextjs_13_to_15.aif


  [      html] 

blocks=1.00  types=1.00  semantic=1.00  meta=0.50  inline=1.00  overall=0.90
  [  markdown] 

blocks=1.93  types=0.00  semantic=1.00  meta=0.50  inline=1.00  overall=0.70
  [      json] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=1.00  overall=1.00
  migration_typescript_strict.aif


  [      html] 

blocks=1.00  types=1.00  semantic=1.00  meta=0.50  inline=0.94  overall=0.89
  [  markdown] 

blocks=2.25  types=0.00  semantic=1.00  meta=0.50  inline=1.00  overall=0.70
  [      json] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=1.00  overall=1.00
  climate_data.aif


  [      html] 

blocks=1.00  types=1.00  semantic=1.00  meta=0.33  inline=1.00  overall=0.87
  [  markdown] 

blocks=1.40  types=0.15  semantic=0.00  meta=0.17  inline=1.00  overall=0.26
  [      json] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=1.00  overall=1.00
  api-debugging.aif


  [      html] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=1.00  overall=1.00
  [  markdown] 

blocks=1.67  types=0.00  semantic=1.00  meta=0.50  inline=1.00  overall=0.70
  [      json] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=1.00  overall=1.00
  base-debugging.aif


  [      html] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=1.00  overall=1.00
  [  markdown] 

blocks=1.33  types=0.00  semantic=1.00  meta=0.50  inline=1.00  overall=0.70
  [      json] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=1.00  overall=1.00
  claude-opus-4-5-migration.aif


  [      html] 

blocks=1.00  types=1.00  semantic=1.00  meta=0.67  inline=1.00  overall=0.93
  [  markdown] 

blocks=2.38  types=0.00  semantic=1.00  meta=0.33  inline=0.93  overall=0.65
  [      json] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=1.00  overall=1.00
  code_review.aif


  [      html] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=1.00  overall=1.00
  [  markdown] 

blocks=2.29  types=0.00  semantic=1.00  meta=0.50  inline=0.89  overall=0.68
  [      json] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=1.00  overall=1.00
  commit-commands.aif


  [      html] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=1.00  overall=1.00
  [  markdown] 

blocks=1.73  types=0.00  semantic=1.00  meta=0.50  inline=0.91  overall=0.68
  [      json] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=1.00  overall=1.00
  feature-dev.aif


  [      html] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=1.00  overall=1.00
  [  markdown] 

blocks=1.38  types=0.00  semantic=1.00  meta=0.50  inline=1.00  overall=0.70
  [      json] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=1.00  overall=1.00
  frontend-design.aif


  [      html] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=1.00  overall=1.00
  [  markdown] 

blocks=1.42  types=0.00  semantic=1.00  meta=0.50  inline=0.86  overall=0.67
  [      json] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=1.00  overall=1.00
  gstack-autoplan.aif


  [      html] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=0.38  overall=0.88
  [  markdown] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=0.83  overall=0.97
  [      json] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=1.00  overall=1.00
  gstack-benchmark.aif


  [      html] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=0.43  overall=0.89
  [  markdown] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=0.73  overall=0.95
  [      json] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=1.00  overall=1.00
  gstack-canary.aif


  [      html] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=0.45  overall=0.89
  [  markdown] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=0.91  overall=0.98
  [      json] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=1.00  overall=1.00
  gstack-careful.aif


  [      html] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=0.63  overall=0.93
  [  markdown] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=0.98  overall=1.00
  [      json] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=1.00  overall=1.00
  gstack-checkpoint.aif


  [      html] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=0.42  overall=0.88
  [  markdown] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=0.83  overall=0.97
  [      json] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=1.00  overall=1.00
  gstack-codex.aif


  [      html] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=0.49  overall=0.90
  [  markdown] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=0.84  overall=0.97
  [      json] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=1.00  overall=1.00
  gstack-freeze.aif


  [      html] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=0.42  overall=0.88
  [  markdown] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=0.93  overall=0.99
  [      json] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=1.00  overall=1.00
  gstack-gstack-upgrade.aif


  [      html] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=0.51  overall=0.90
  [  markdown] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=0.70  overall=0.94
  [      json] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=1.00  overall=1.00
  gstack-guard.aif


  [      html] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=0.44  overall=0.89
  [  markdown] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=0.94  overall=0.99
  [      json] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=1.00  overall=1.00
  gstack-investigate.aif


  [      html] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=0.44  overall=0.89
  [  markdown] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=0.88  overall=0.98
  [      json] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=1.00  overall=1.00
  gstack-learn.aif


  [      html] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=0.38  overall=0.88
  [  markdown] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=0.88  overall=0.98
  [      json] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=1.00  overall=1.00
  gstack-qa.aif


  [      html] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=0.49  overall=0.90
  [  markdown] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=0.87  overall=0.97
  [      json] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=1.00  overall=1.00
  gstack-review.aif


  [      html] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=0.45  overall=0.89
  [  markdown] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=0.89  overall=0.98
  [      json] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=1.00  overall=1.00
  gstack-ship.aif


  [      html] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=0.49  overall=0.90
  [  markdown] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=0.40  overall=0.88
  [      json] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=1.00  overall=1.00
  gstack-unfreeze.aif


  [      html] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=0.48  overall=0.90
  [  markdown] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=0.87  overall=0.97
  [      json] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=1.00  overall=1.00
  security-guidance.aif


  [      html] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=1.00  overall=1.00
  [  markdown] 

blocks=2.00  types=0.00  semantic=1.00  meta=0.50  inline=0.79  overall=0.66
  [      json] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=1.00  overall=1.00
  superpowers-brainstorming.aif


  [      html] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=0.46  overall=0.89
  [  markdown] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=0.99  overall=1.00
  [      json] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=1.00  overall=1.00
  superpowers-dispatching-parallel-agents.aif


  [      html] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=0.49  overall=0.90
  [  markdown] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=0.94  overall=0.99
  [      json] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=1.00  overall=1.00
  superpowers-executing-plans.aif


  [      html] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=0.35  overall=0.87
  [  markdown] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=0.99  overall=1.00
  [      json] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=1.00  overall=1.00
  superpowers-finishing-a-development-branch.aif


  [      html] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=0.50  overall=0.90
  [  markdown] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=0.91  overall=0.98
  [      json] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=1.00  overall=1.00
  superpowers-receiving-code-review.aif


  [      html] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=0.48  overall=0.90
  [  markdown] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=0.88  overall=0.98
  [      json] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=1.00  overall=1.00
  superpowers-requesting-code-review.aif


  [      html] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=0.48  overall=0.90
  [  markdown] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=0.93  overall=0.99
  [      json] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=1.00  overall=1.00
  superpowers-subagent-driven-development.aif


  [      html] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=0.46  overall=0.89
  [  markdown] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=0.97  overall=0.99
  [      json] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=1.00  overall=1.00
  superpowers-using-git-worktrees.aif


  [      html] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=0.47  overall=0.89
  [  markdown] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=0.83  overall=0.97
  [      json] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=1.00  overall=1.00
  superpowers-using-superpowers.aif


  [      html] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=0.34  overall=0.87
  [  markdown] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=0.99  overall=1.00
  [      json] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=1.00  overall=1.00
  superpowers-verification-before-completion.aif


  [      html] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=0.34  overall=0.87
  [  markdown] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=0.91  overall=0.98
  [      json] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=1.00  overall=1.00
  superpowers-writing-plans.aif


  [      html] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=0.44  overall=0.89
  [  markdown] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=0.89  overall=0.98
  [      json] 

blocks=1.00  types=1.00  semantic=1.00  meta=1.00  inline=1.00  overall=1.00

Completed: 120 roundtrip tests


In [4]:
# ── Summary by Format ─────────────────────────────────────────────────────────

good = [r for r in results if "error" not in r]
errors = [r for r in results if "error" in r]

print(f"{'='*80}")
print(f"  SUMMARY BY FORMAT")
print(f"{'='*80}")

header = f"{'Format':<12} {'Overall Fidelity':>16} {'Block Types':>12} {'Semantic Types':>15} {'Metadata':>10} {'Inlines':>10}"
print(header)
print("-" * len(header))

format_summaries = {}
for fmt in FORMATS:
    fmt_results = [r for r in good if r["format"] == fmt]
    if not fmt_results:
        print(f"{fmt:<12} {'(no data)':>16}")
        continue
    n = len(fmt_results)
    avg = {
        "overall": sum(r["overall_fidelity"] for r in fmt_results) / n,
        "block_types": sum(r["block_type_preservation"] for r in fmt_results) / n,
        "semantic": sum(r["semantic_type_preservation"] for r in fmt_results) / n,
        "metadata": sum(r["metadata_preservation"] for r in fmt_results) / n,
        "inlines": sum(r["inline_fidelity"] for r in fmt_results) / n,
        "count": n,
    }
    format_summaries[fmt] = avg
    print(f"{fmt:<12} {avg['overall']:>16.4f} {avg['block_types']:>12.4f} {avg['semantic']:>15.4f} {avg['metadata']:>10.4f} {avg['inlines']:>10.4f}")

if errors:
    print(f"\nFailed roundtrips ({len(errors)}):")
    for e in errors:
        print(f"  {e['file']} [{e['format']}]: {e['error']}")

  SUMMARY BY FORMAT
Format       Overall Fidelity  Block Types  Semantic Types   Metadata    Inlines
--------------------------------------------------------------------------------
html                   0.9124       1.0000          1.0000     0.9292     0.6331
markdown               0.8443       0.6615          0.9250     0.8083     0.9016
json                   1.0000       1.0000          1.0000     1.0000     1.0000


In [5]:
# ── Per-File Results ──────────────────────────────────────────────────────────

print(f"{'='*100}")
print(f"  DETAILED PER-FILE RESULTS")
print(f"{'='*100}")

header = f"{'File':<38} {'Format':<10} {'Blocks':>7} {'Types':>7} {'Semntic':>7} {'Meta':>7} {'Inline':>7} {'Overall':>7}"
print(header)
print("-" * len(header))

for r in good:
    print(f"{r['file']:<38} {r['format']:<10} "
          f"{r['block_count_ratio']:>7.2f} "
          f"{r['block_type_preservation']:>7.2f} "
          f"{r['semantic_type_preservation']:>7.2f} "
          f"{r['metadata_preservation']:>7.2f} "
          f"{r['inline_fidelity']:>7.2f} "
          f"{r['overall_fidelity']:>7.2f}")

  DETAILED PER-FILE RESULTS
File                                   Format      Blocks   Types Semntic    Meta  Inline Overall
-------------------------------------------------------------------------------------------------
simple.aif                             html          1.00    1.00    1.00    1.00    1.00    1.00
simple.aif                             markdown      1.27    0.09    0.00    0.50    1.00    0.32
simple.aif                             json          1.00    1.00    1.00    1.00    1.00    1.00
wiki_article.aif                       html          1.00    1.00    1.00    0.67    1.00    0.93
wiki_article.aif                       markdown      1.30    0.22    0.00    0.33    1.00    0.31
wiki_article.aif                       json          1.00    1.00    1.00    1.00    1.00    1.00
migration_eslint_flat_config.aif       html          1.00    1.00    1.00    0.50    0.67    0.83
migration_eslint_flat_config.aif       markdown      1.93    0.00    1.00    0.50    0.97 

In [6]:
# ── Save results ─────────────────────────────────────────────────────────────

output_path = PROJECT_ROOT / "benchmarks" / "roundtrip" / "results.json"
with open(output_path, "w") as f:
    json.dump(results, f, indent=2)
print(f"Results saved to {output_path}")

Results saved to /Users/liqunchen/Claude_Project/token_efficient/benchmarks/roundtrip/results.json


## Findings

### Format Fidelity Tiers

| Format | Overall Fidelity | Interpretation |
|--------|:----------------:|----------------|
| **JSON** | **1.00** | Lossless -- ideal for machine-to-machine exchange |
| **HTML** | **0.91** | Very high -- AIF CSS classes (`aif-claim`, `aif-evidence`, etc.) enable near-perfect reconstruction |
| **Markdown** | **0.84** | Moderate -- Markdown preserves basic formatting but loses semantic block types |

### Semantic Structure Is What Makes LLMs Understand Documents

Semantic type preservation is the metric that justifies AIF's existence. When a document roundtrips
through Markdown, `@claim`, `@evidence`, and `@definition` blocks degrade into plain paragraphs or
blockquotes. The LLM loses the ability to distinguish a factual claim from supporting evidence --
they all look like text.

This is why the metric is **double-weighted** in the overall score: losing structure means the
LLM cannot identify what kind of content it is reading, which directly impacts its ability to
follow instructions and reason accurately.

### Key Observations

- **JSON roundtrip is perfectly lossless** (1.00 across all metrics, all 40 files). This validates
  the AST serialization layer.
- **HTML achieves 0.91** because AIF's HTML emitter annotates every semantic block with CSS classes
  (`class="aif-claim"`, `class="aif-evidence"`), which the importer detects and reconstructs.
- **Markdown scores 0.84** overall -- basic formatting survives but semantic block types are lost.
  A `@claim` becomes a `> blockquote` or flat paragraph with no semantic type annotation.
- **Typed blocks are the difference between "text" and "structured input."** When an LLM receives
  `@claim[id=c1, refs=e1]`, it knows this is a factual assertion linked to evidence. When it receives
  a paragraph in Markdown, it must *infer* the content's role -- and inference means guessing.
  For agent workflows, RAG with typed chunks, and compliance checking, this distinction is critical.